File: nbs_customization/public/js/payment_entry.js
// Copyright (c) 2026, NBS Solutions and contributors
// For license information, please see license.txt

frappe.provide("erpnext.accounts.pos");

// Route cheque modes of payment to the direction-correct clearing account when
// the base code fetches the mode's payment account (the stock account pickers
// cannot list clearing accounts, which have no Bank/Cash account type).
(function () {
	let pos_account_getter = erpnext.accounts.pos.get_payment_mode_account;

	erpnext.accounts.pos.get_payment_mode_account = function (frm, mode_of_payment, callback) {
		if (!mode_of_payment) {
			if (pos_account_getter) return pos_account_getter(frm, mode_of_payment, callback);
			return null;
		}
		frappe.db.get_value(
			"Mode of Payment",
			mode_of_payment,
			["is_check", "clearing_account_inward", "clearing_account_outward"],
			(r) => {
				let account = "";
				if (r && r.is_check) {
					account =
						frm.doc.payment_type === "Receive"
							? r.clearing_account_inward
							: r.clearing_account_outward;
				}
				if (account) {
					callback(account);
				} else if (pos_account_getter) {
					pos_account_getter(frm, mode_of_payment, callback);
				}
			},
		);
	};
})();

function maybe_add_receipt_button(frm) {
	if (
		frm.doc.docstatus !== 1 ||
		frm.doc.payment_type !== "Receive" ||
		frm.doc.party_type !== "Customer"
	)
		return;

	if (frm.doc.custom_receipt) {
		frm.add_custom_button(
			__("Receipt"),
			() => frappe.set_route("Form", "Receipt", frm.doc.custom_receipt),
			__("View"),
		);
	} else {
		frm.add_custom_button(
			__("Receipt"),
			() =>
				frappe.model.open_mapped_doc({
					method: "nbs_customization.nbs_customization.doctype.receipt.receipt.create_receipt_from_pe",
					frm: frm,
				}),
			__("Create"),
		);
		frm.page.set_inner_btn_group_as_primary(__("Create"));
	}
}

function add_check_clearing_buttons(frm) {
	if (
		frm.doc.docstatus !== 1 ||
		!frm.doc.is_check ||
		frm.doc.check_cleared ||
		frm.doc.check_returned ||
		frm.doc.clearing_journal_entry
	)
		return;

	frm.add_custom_button(__("Mark Check Cleared"), () => clearing_dialog(frm), __("Cheque"));

	frm.add_custom_button(
		__("Mark Check Returned"),
		() => {
			frappe.confirm(
				__(
					"Mark this cheque as returned/bounced? This reverses any clearing and cancels the Payment Entry, re-opening the allocated invoices.",
				),
				() => {
					frappe.call({
						method: "nbs_customization.controllers.payment_entry.mark_check_returned",
						args: { name: frm.doc.name },
						freeze: true,
						freeze_message: __("Marking cheque as returned..."),
						callback(r) {
							if (!r.exc) {
								frappe.msgprint(
									__(
										"Payment Entry {0} cancelled. Invoice allocations are re-opened.",
										[frm.doc.name],
									),
								);
								frm.reload_doc();
							}
						},
					});
				},
			);
		},
		__("Cheque"),
	);
}

function clearing_dialog(frm) {
	let d = new frappe.ui.Dialog({
		title: __("Mark Check Cleared"),
		fields: [
			{
				fieldtype: "Link",
				fieldname: "clearing_destination_account",
				label: __("Destination Account"),
				options: "Account",
				default: frm.doc.clearing_destination_account,
				reqd: 1,
				get_query() {
					return {
						filters: {
							company: frm.doc.company,
							is_group: 0,
							account_type: ["in", ["Bank", "Cash"]],
						},
					};
				},
			},
			{
				fieldtype: "Date",
				fieldname: "clearing_date",
				label: __("Clearing Date"),
				default: frappe.datetime.get_today(),
				reqd: 1,
			},
		],
		primary_action_label: __("Clear"),
		primary_action(values) {
			frappe.call({
				method: "nbs_customization.controllers.payment_entry.mark_check_cleared",
				args: {
					name: frm.doc.name,
					destination_account: values.clearing_destination_account,
					clearing_date: values.clearing_date,
				},
				freeze: true,
				freeze_message: __("Clearing cheque..."),
				callback(r) {
					if (!r.exc) {
						frappe.msgprint(
							__("Cheque cleared. Journal Entry {0} created.", [
								r.message.journal_entry,
							]),
						);
						d.hide();
						frm.reload_doc();
					}
				},
			});
		},
	});
	d.show();
}

frappe.ui.form.on("Payment Entry", {
	refresh(frm) {
		maybe_add_receipt_button(frm);
		add_check_clearing_buttons(frm);
	},
});

File: nbs_customization/controllers/payment_entry.py
# Copyright (c) 2026, Charles Byakutaga/NBS and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import today

from nbs_customization.controllers.check_clearing import (
	create_check_clearing_je,
	get_check_mop,
	stamp_check_cleared,
)


def validate_check_payment_entry(pe, method=None):
	"""Force cheque Payment Entries onto the direction-correct clearing account."""
	if pe.payment_type not in ("Receive", "Pay") or not pe.mode_of_payment:
		return
	mop = get_check_mop(pe.mode_of_payment)
	if not mop.get("is_check"):
		return
	expected = (
		mop.get("clearing_account_inward")
		if pe.payment_type == "Receive"
		else mop.get("clearing_account_outward")
	)
	if not expected:
		frappe.throw(
			_("Mode of Payment '{0}' is a cheque mode but has no {1} clearing account set.").format(
				pe.mode_of_payment, "inward" if pe.payment_type == "Receive" else "outward"
			)
		)
	pe.set("is_check", 1)
	bank_side = "paid_to" if pe.payment_type == "Receive" else "paid_from"
	if pe.get(bank_side) != expected:
		pe.set(bank_side, expected)
	if not pe.clearing_destination_account and mop.get("default_clearing_destination"):
		pe.clearing_destination_account = mop.get("default_clearing_destination")


@frappe.whitelist(methods=["POST"])
def mark_check_cleared(name: str, destination_account: str | None = None, clearing_date: str | None = None):
	"""Button-driven clearing of a cheque Payment Entry.

	The funds land in `destination_account` (Bank or Cash); defaults to the
	mode of payment default or the company default bank account.
	"""
	pe = frappe.get_doc("Payment Entry", name)
	if pe.docstatus != 1:
		frappe.throw(_("Only submitted Payment Entries can be marked cleared."))
	mop = get_check_mop(pe.mode_of_payment)
	if not mop.get("is_check"):
		frappe.throw(_("{0} is not a cheque Payment Entry.").format(pe.name))
	if pe.check_cleared or pe.check_returned:
		frappe.throw(_("{0} is already cleared or returned.").format(pe.name))

	chosen = destination_account or pe.clearing_destination_account or mop.get("default_clearing_destination")
	if not chosen:
		chosen = frappe.get_cached_value("Company", pe.company, "default_bank_account")
	clear_date = clearing_date or today()

	je_name = create_check_clearing_je(pe, chosen, clear_date)
	rows = stamp_check_cleared(pe, je_name, clear_date, "Button", chosen)
	if rows != 1:
		# lost a race with another clear; undo the JE just created
		frappe.get_doc("Journal Entry", je_name).cancel()
		frappe.throw(_("{0} was already cleared.").format(pe.name))
	return {"journal_entry": je_name, "cleared": True}


@frappe.whitelist(methods=["POST"])
def mark_check_returned(name: str):
	"""A cheque bounced/returned.

	Reverses any prior clearing JE, then cancels the Payment Entry itself so the
	native cancel flow re-opens invoice allocations and reverses the GL.
	"""
	pe = frappe.get_doc("Payment Entry", name)
	if pe.docstatus != 1:
		frappe.throw(_("Only submitted Payment Entries can be marked returned."))
	if pe.check_returned:
		frappe.throw(_("{0} is already marked returned.").format(pe.name))

	if pe.clearing_journal_entry:
		# Break the circular PE <-> JE link first, otherwise Frappe blocks the
		# clearing Journal Entry from being cancelled (the submitted Payment
		# Entry points back at it via `clearing_journal_entry`).
		frappe.db.set_value("Payment Entry", pe.name, "clearing_journal_entry", None, update_modified=False)
		clearing_je = frappe.get_doc("Journal Entry", pe.clearing_journal_entry)
		if clearing_je.docstatus == 1:
			clearing_je.cancel()

	pe.cancel()
	frappe.db.set_value(
		"Payment Entry",
		pe.name,
		{
			"check_returned": 1,
			"check_return_date": today(),
			"check_cleared": 0,
			"check_clearing_date": None,
			"check_cleared_source": None,
			"clearing_destination_account": None,
			"clearance_date": None,
		},
	)
	return {"returned": True}
